### ISSUE TITLE
$q.resolve() doesn't run digest UNLESS there is a .then()

### ISSUE THREAD

**Issue by oodavid** [angular/angular.js#14339]
**Feature or bug?**

Bug

**Current behavior?**

I want to wrap async code in $q.resolve() in order to trigger a digest cycle and update the scope.

However, this only happens IF $q.resolve() has further methods, such as .then().

**Plunkr Demo**

https://plnkr.co/edit/q7eVs9?p=preview

**Expected behavior?**

I expect the scope to update regardless of whether $q.resolve() has further methods.

**What is the motivation / use case for changing the behavior?**

Adding this type of code (below) is redundant, breaks expected behaviour and is likely to cause unexpected errors when omitted.

```
$q.resolve(
  // ...
)
.then(function(){
  // Doesn't do anything.
  // Angular just needs a $q.then() in order to update the scope.
  // Possibly a bug?
});
```

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**

**Comment by Narretz** [comment]
I think that's for performance, because a promise without a then doesn't really make sense. I'm also a bit puzzled about your way of using $q.resolve() in the plunkr. Is this just for illustraton or is this your actual implementation?

**Comment by Narretz** [comment]
@lgalfaso might know more about this

**Comment by lgalfaso** [comment]
TL;DR; This works as expected.

If you use a non-angular aware library (and this is unrelated to `$q`) then when you change the scope you have to do so inside an `$apply` of call `$digest`. Here is the same example fixed.
https://plnkr.co/edit/KeEaYO3RsLsAeiNYtuOv?p=preview
The reason that adding the extra `.then` worked is because by doing so, when the promise is resolved, would trigger a digest for you so it would run your `noop` function inside a `$digest`.

BTW, if you want to use Bluebird, then there is a mechanism so Bluebird will call `$digest` for you and it is also possible to replace `$q` with Bluebird. Both options should be easy to find.

**Comment by gkalpak** [comment]
I think this is (kind of) expected.

--
`$q.resolve()` _does_ invoke a digest. But it does so right after it is called (unless one is already in progress).

The "problem" is that nobody notifies Angular when the native promise is resolved, 2 seconds later (thus no digest). If you chain the promise using `.then()`, then you basically get a new `$q` promise (which _is_ Angular aware and _does_ invoke a digest once resolved), It just so happens (by design actually) that this new `$q`-based promise gets resolved _after_ the first, native promise is resolved.

--
If you used `$q` instead of the native promise, everything would work as expected, even without the extra `.then()`. (Note: `$q` can also be used as a constructor (`new $q(function (resolve, reject) {...})`).
[Updated plnkr][1]

[1]: https://plnkr.co/edit/uGp2FE6M0XnnwboLagVJ?p=preview
